### ISSUE TITLE
reduce peak memory usage during training via pytorch caching allocator "preallocation" strategy

### ISSUE THREAD

**Issue by tmm1** [huggingface/transformers#25831]
### Feature request

In the PyTorch tuning guide, there's a section on preallocation as a technique to reduce memory usage when dealing with variable length inputs.

https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html#preallocate-memory-in-case-of-variable-input-length

<img width="544" alt="image" src="https://github.com/huggingface/transformers/assets/2567/d4200505-ae08-46f4-a245-f13b4b9b9646">

It would be great to implement this in transformers.Trainer as an option

### Motivation

In investigating memory usage for training, I discovered pytorch memory fragmentation can be a big problem.

One workaround is to pad inputs to the same length, i.e. https://github.com/OpenAccess-AI-Collective/axolotl/pull/498#issuecomment-1696539038

However it would be better to not have to use padding at all just to control memory usage. The preallocation strategy is a better approach, and should only require 2-3 passes with random inputs.

### Your contribution

I can work on a PR for this if the idea is acceptable. It would also be helpful to decide on the name of the configuration option.

**Comment by ydshieh** [comment]
It looks like we don't really to change anything in trainer, but just make a forward/backward pass with a first (max sized) input batch, then zero out the gradients. Everything could be done outside the trainer by the user.

**Comment by tmm1** [comment]
> Everything could be done outside the trainer by the user.

This may be true, but also the Trainer implementation could be using all sorts of extra features like apex/amp/loss_context_manager/grad_scaling etc. These could all create extra tensors that are based on the input shape, so I think it would be best to build this into the Trainer itself.

**Comment by ydshieh** [comment]
The `trainer` is a file of 4k lines - we really want to keep it not to continue to grow 🙏

**Comment by tmm1** [comment]
I tried to do this outside the trainer but it is not effective.

After three steps of forward+backward, the torch cache size is only 2.4GB.

But when I measure it after the first trainer step, its 3-4x higher.

So the trainer is doing a lot more work. As you said there are 4k lines, so there is a lot of complexity. It's not easy to replicate that outside the trainer for every user.

If there isn't interest in this feature, I guess most users should consider padding inputs. In my test with llama2-7b + lora, peak memory usage could be 25% higher with variable length inputs.
